### PR TITLE
1537/authz policies

### LINKED ISSUE #1537: Use auth and authz policies also with the /auth endpoint
Several auth and authz policies are not used with the webui. 
It is because those policies are decorators to ``/validate/check`` but of course can not easily be decorators to the ``/auth`` endpoint.

We need to investigate, how it is possible to also honour the policies in the web UI **if** ``loginmode=privacyIDEA`` is set.

### PR DESCRIPTION
**Pull request by cornelinux**
I finished the orphaned branch of @fredreichbier.

Closes #1537 

I added some more explanation to the implementation for the max_auth_success and max_auth_fail policies in the comment to the issue #1537.

Had to resolve some conflicts, I hope I did it in the right way.

@plettich Maybe you can especially look into if everything with JSON in the request is done right according to the current state of the master. (The original branch was already older than 2 months)

### CODE REVIEW

File: tests/test_api_roles.py [hunk 1]
```diff
@@ -618,6 +620,338 @@ def test_08_user_can_set_pin(self):
...
+    def test_09_authz_user_detail(self):
+        # Test behavior of ADDUSERINRESPONSE and ADDRESOLVERINRESPONSE policy action for /auth endpoint
+        # Normally, no userinfo and realm/resolver are found in the response
+        with self.app.test_request_context('/auth',
+                                           method='POST',
+                                           data={"username": "selfservice@realm1",
+                                                 "password": "test"}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            content = json.loads(res.data.decode('utf8'))
```
Comment (plettich) [inline tests/test_api_roles.py]: I'll change the json stuff and push it here.

File: tests/test_api_roles.py [hunk 2]
```diff
@@ -618,6 +620,338 @@ def test_08_user_can_set_pin(self):
...
+                                           data={"username": "selfservice@realm1",
+                                                 "password": "somepin"}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(res.status_code, 200)
+            content = json.loads(res.data.decode('utf8'))
+            self.assertNotIn("detail", content)
+            result = content.get("result")
+            self.assertTrue(result.get("status"), res.data)
+
+        # Authentication still works for internal admins
```
Comment (plettich) [inline tests/test_api_roles.py]: I think this test is not necessary here.

File: tests/testdata/tmp_directory [hunk 3]
```diff
@@ -1 +1 @@
...s': {'oid': '2', 'cn': 'alice', 'accountExpires': 131024988000000000, 'homeDirectory': '/home/alice', 'mobile': ['1234', '45678'], 'userPassword': 'alicepw', 'givenName': 'Alice', 'objectGUID': '\xef6\x9b\x03\xc0\xe7\xf3B\x9b\xf9\xcajl\rM1', 'email': 'alice@test.com', 'sn': 'Cooper'}}, {'dn': 'cn=bob,ou=example,o=test', 'attributes': {'oid': '3', 'cn': 'bob', 'accountExpires': 9223372036854775807, 'homeDirectory': '/home/bob', 'mobile': '123456', 'userPassword': 'bobpw\xc3\xa9\xc3\xa9\xc3\xa9', 'givenName': 'Robert', 'objectGUID': '\xef6\x9b\x03\xc0\xe7\xf3B\x9b\xf9\xcajl\rMw', 'email': 'bob@example.com', 'sn': 'Marley'}}, {'dn': 'cn=manager,ou=example,o=test', 'attributes': {'oid': '1', 'cn': 'manager', 'accountExpires': 9223372036854775807, 'mobile': '123354', 'userPassword': 'ldaptest', 'givenName': 'Corny', 'objectGUID': '\xef6\x9b\x03\xc0\xe7\xf3B\x9b\xf9\xcajl\rMT', 'email': 'ck@o', 'sn': 'keule'}}, {'dn': 'cn=frank,ou=sales,o=test', 'attributes': {'oid': '5', 'cn': 'frank', 'accountExpires': 9223372036854775807, 'mobile': '123354', 'userPassword': 'ldaptest', 'givenName': 'Frank', 'objectGUID': '\xef7\x9b\x03\xc0\xe7\xf3B\x9b\xf9\xcajl\rMT', 'email': 'fh@o', 'sn': 'Hause'}}]
```
Comment (plettich) [inline tests/testdata/tmp_directory]: I think, this is a remnant of the test process. I'll remove it in my commit.

File: privacyidea/api/auth.py [hunk 4]
```diff
@@ -264,19 +272,22 @@ def get_auth_token():
...
         if role == ROLE.ADMIN:
             g.audit_object.log({"user": "",
                                 "administrator": user_obj.login,
                                 "realm": user_obj.realm,
                                 "resolver": user_obj.resolver,
-                                "serial": details.get('serial', None) if details else None,
-                                "info": log_used_user(user_obj)})
+                                "serial": details.get('serial', None),
+                                "info": u"{0!s}|loginmode={1!s}".format(log_used_user(user_obj),
+                                        details.get("loginmode"))})
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/auth.py]: I dont' get it.
In line 275 there is

    details = details or {}

So details will be a dict.
If the dict does not contain the key "loginmode" then ``details.get("loginmode")`` will return None anyways.

   details.get("lloginode") == details.get("loginmode", None)

...or what am I missing?

File: privacyidea/api/auth.py [hunk 5]
```diff
@@ -264,19 +272,22 @@ def get_auth_token():
...
         else:
             g.audit_object.log({"user": user_obj.login,
                                 "realm": user_obj.realm,
                                 "resolver": user_obj.resolver,
-                                "serial": details.get('serial', None) if details else None,
-                                "info": log_used_user(user_obj)})
+                                "serial": details.get('serial', None),
+                                "info": u"{0!s}|loginmode={1!s}".format(log_used_user(user_obj),
+                                        details.get("loginmode"))})
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/auth.py]: s.o.

File: privacyidea/api/lib/postpolicy.py [hunk 6]
```diff
@@ -303,7 +303,9 @@ def no_detail_on_success(request, response):
     if detailPol and content.get("result", {}).get("value"):
         # The policy was set, we need to strip the details, if the
         # authentication was successful. (value=true)
-        del content["detail"]
+        # But only if there is a "detail" key in the response.
+        if "detail" in content:
+            del content["detail"]
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/lib/postpolicy.py]: Sounds, cool. Will do.

File: privacyidea/api/lib/postpolicy.py [hunk 7]
```diff
@@ -328,20 +330,21 @@ def add_user_detail_to_response(request, response):
...
         #  details
-        ui = request.User.info
+        ui = request.User.info.copy()
         ui["password"] = ""
         for key, value in ui.items():
             if type(value) == datetime.datetime:
                 ui[key] = str(value)
-        content["detail"]["user"] = ui
+        content.setdefault("detail", {})["user"] = ui
+        response.data = json.dumps(content)
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/lib/postpolicy.py]: Will do

File: tests/test_api_roles.py (hunk 1 above)
Comment (cornelinux) [reply tests/test_api_roles.py]: thx

File: tests/testdata/tmp_directory (hunk 3 above)
Comment (cornelinux) [reply tests/testdata/tmp_directory]: thx,

File: privacyidea/api/lib/postpolicy.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/lib/postpolicy.py]: done in ffa0b90f

File: privacyidea/api/lib/postpolicy.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/api/lib/postpolicy.py]: done in  ffa0b90f

File: privacyidea/api/auth.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (plettich) [reply privacyidea/api/auth.py]: My bad ... You are right. I got confused by the `details.get('serial', None)` above.

**Pull request merged**